### PR TITLE
SW-176 add grbl latest version to mr beam plugin

### PR DESCRIPTION
**Pull request by JFSolution**
update of https://github.com/mrbeam/MrBeamPlugin/pull/1352

this will fix that new images won't update grbl

### CODE REVIEW

**Review by khaledsherkawi — DISMISSED** [review]

**Comment by khaledsherkawi** [comment]
@Josef-MrBeam I asked Axel to have a quick look as well. So please don't merge without his approval.

File: octoprint_mrbeam/migrate.py
Comment (amrsoll) [inline octoprint_mrbeam/migrate.py]: use `settings.remove("grbl_version_lastknown")`

File: octoprint_mrbeam/migrate.py
Comment (amrsoll) [inline octoprint_mrbeam/migrate.py]: you are using `self.plugin._settings` a lot. I would recommend to assign it to a variable for this code here, like `settings = self.plugin._settings`

This can also help fit the lines inside the 80 character limit.

File: octoprint_mrbeam/printing/comm_acc2.py
Comment (amrsoll) [inline octoprint_mrbeam/printing/comm_acc2.py]: Same here

`settings = self._mrbeam_plugin_implementation._settings`

Although accessing the global variable `_mrbeam_plugin_implementation` is Not recommended.
To fix that I recommend to move this function out of the `comm_acc2` module. Responsability-wise, this is up to the `Laser` object to ask the `Laser.disconnect()`, which closes the `_comm` object, then run that shell command. After that you can try to re-`connect()` from the `Laser` class.

The problem is that the `Laser` class only has access to the "printer profile" or - in our case - "laser profile". And this makes sense in case we have the same general OctoPrint settings in the plugin, but we want to adapt it to different machines, then this config is the place where we isolate the changes.

Hopefully you can manage to migrate this responsability to the `Laser` class.

Please keep the grbl settings in the laser cutter profile. This can require quite some migration such as `grbl_version_lastknown` into that profile too.

File: octoprint_mrbeam/printing/comm_acc2.py
Comment (amrsoll) [inline octoprint_mrbeam/printing/comm_acc2.py]: Please just remove useless exceptions

**Review by amrsoll — CHANGES_REQUESTED** [review]
After I FINALLY understood what went wrong (whoops :sweat_smile: ), I think I found the problem! The laser cutter profile could not load properly because `LaserCutterProfile.get(identifier)` could not find the `identifier` in the `LASER_PROFILE_IDENTIFIERS` - which don't include the latest identifiers, namely `"MrBeam2S"` and `"MrBeam2R"`. You can add those profiles in the `octoprint_mrbeam/printing/profiles/`.

After that , I thiiink the GRBL auto-updates should work. But I believe my comments on refactoring still apply.

**Comment by amrsoll** [comment]
Actually, if you have a look at `octoprint.printer.profile.PrinterProfileManager.save()`, then you see that you can ask it to fire an Event. You can subscribe to that Event in the `Laser` class. If the grbl version of the current profile doesn't match the installed one, then you can start flashing the new grbl version.

This allows us to 
- remove all grbl stuff from the `__init__` file, 
- remove `reset_grbl_auto_update_config` because we compare versions to what we expect
- `auto_update_grbl()` only needs to know what the new version is. No need to compare to anything else etc...

File: octoprint_mrbeam/migrate.py
Comment (JFSolution) [reply octoprint_mrbeam/migrate.py]: yes, good point

File: octoprint_mrbeam/migrate.py
Comment (JFSolution) [reply octoprint_mrbeam/migrate.py]: this will only save 13 signs :), so I think it is not urgend, but yes I can do it :)

File: octoprint_mrbeam/printing/comm_acc2.py
Comment (JFSolution) [reply octoprint_mrbeam/printing/comm_acc2.py]: I didn't add it, and I think there is a reason why it is there, so I just changed the message

**Comment by JFSolution** [comment]
> After I FINALLY understood what went wrong (whoops 😅 ), I think I found the problem! The laser cutter profile could not load properly because `LaserCutterProfile.get(identifier)` could not find the `identifier` in the `LASER_PROFILE_IDENTIFIERS` - which don't include the latest identifiers, namely `"MrBeam2S"` and `"MrBeam2R"`. You can add those profiles in the `octoprint_mrbeam/printing/profiles/`.
> 
> After that , I thiiink the GRBL auto-updates should work. But I believe my comments on refactoring still apply.

Nop, that won't fix it, because we removed the profile file thing, it will work now completly without a profile file, it only uses the coded profile and it seems like there is no way to add anything there while it is running

File: octoprint_mrbeam/printing/comm_acc2.py
Comment (amrsoll) [reply octoprint_mrbeam/printing/comm_acc2.py]: There are only bad reasons for having the exception exactly here. As a rule :
1. Only catch the exceptions you can foresee
2. If exceptions don't escalate or get logged (in the example of threads for example) - then you can have the luxury of just logging all the exceptions.

In every case where you catch every exception, you Should raise it again.

I know it is not your code originally, but when we touch some function, we want to leave it in a better state than we started with. Therefore we should also make sure to refactor bad practices.

**Comment by amrsoll** [comment]
Yes and no. It will still use the files over the coded defaults. If you have a file with the correct ID, then it will overwrite the default settings.
That is why you can still save the config and it should override the hardcoded configs.

File: octoprint_mrbeam/__init__.py
Comment (ahmed-mrbeam) [inline octoprint_mrbeam/__init__.py:2677]: is this format change only ?

**Comment by ahmed-mrbeam** [comment]
@Josef-MrBeam  is there a reason why we deleted all these GRBL files ? is this only cleanup ?

File: octoprint_mrbeam/migrate.py
Comment (ahmed-mrbeam) [inline octoprint_mrbeam/migrate.py]: what if laserCutterProfile is None ??

File: octoprint_mrbeam/printing/comm_acc2.py
Comment (ahmed-mrbeam) [inline octoprint_mrbeam/printing/comm_acc2.py:1829]: That is very broad, are there specific set of exceptions we anticipate ?

File: octoprint_mrbeam/printing/profile.py
Comment (ahmed-mrbeam) [inline octoprint_mrbeam/printing/profile.py:200]: why this if statement is inside the else statement ?, why not on the same level as elif ?

File: octoprint_mrbeam/printing/profile.py
Comment (ahmed-mrbeam) [inline octoprint_mrbeam/printing/profile.py]: more explanation in a code comment please, why we changed the old check completely and using regex now ?

File: octoprint_mrbeam/util/device_info.py
Comment (ahmed-mrbeam) [inline octoprint_mrbeam/util/device_info.py:62]: type is ?.... laser head model ?

Comment (JFSolution) [reply octoprint_mrbeam/util/device_info.py:62]: its the device type for example `MrBeam2S`

File: octoprint_mrbeam/__init__.py
Comment (JFSolution) [reply octoprint_mrbeam/__init__.py:2677]: yes

File: octoprint_mrbeam/printing/profile.py
Comment (JFSolution) [reply octoprint_mrbeam/printing/profile.py:200]: because it would use the same return, so I put them together so it wont't have a copy of it

File: octoprint_mrbeam/printing/profile.py
Comment (JFSolution) [reply octoprint_mrbeam/printing/profile.py]: ok I can add this

File: octoprint_mrbeam/migrate.py
Comment (ahmed-mrbeam) [reply octoprint_mrbeam/migrate.py]: @Josef-MrBeam  this one is still not answered

File: octoprint_mrbeam/printing/comm_acc2.py
Comment (ahmed-mrbeam) [reply octoprint_mrbeam/printing/comm_acc2.py:1829]: @Josef-MrBeam

Comment (JFSolution) [reply octoprint_mrbeam/printing/comm_acc2.py:1829]: As discusses today we have to figure out how we deal with such broad exceptions

**Review by ahmed-mrbeam — DISMISSED** [review]

**Comment by amrsoll** [comment]
All good except for 2 things:
- Please save the dictionary diff between the default and the changed config - This way we don't overwrite the default config if we decide to modify the default values, unless we specifically decide to save these default profile values in the yaml file
- Please check with @ahmed-mrbeam if it's ok to downgrade the version from 0.10.x to 0.9.x

File: octoprint_mrbeam/printing/profile.py
Comment (amrsoll) [inline octoprint_mrbeam/printing/profile.py:271]: Alright, now please add docstring + what changed compared to the overloaded function

**Review by amrsoll — APPROVED** [review]

**Pull request merged**
